File: apartment-portal-BE/src/main/java/com/mytech/apartment/portal/models/Invoice.java
package com.mytech.apartment.portal.models;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Set;
import com.mytech.apartment.portal.models.enums.InvoiceStatus;

@Entity
@Table(name = "invoices")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Invoice {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "apartment_id", nullable = false)
    private Long apartmentId;

    @Column(name = "billing_period", nullable = false)
    private String billingPeriod; // format 'YYYY-MM'

    @Column(name = "issue_date", nullable = false)
    private LocalDate issueDate;

    @Column(name = "due_date", nullable = false)
    private LocalDate dueDate;

    @Column(name = "total_amount", nullable = false)
    private Double totalAmount;

    @Column(nullable = false)
    private InvoiceStatus status;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @CreationTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "invoice", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Set<InvoiceItem> items;

    @OneToMany(mappedBy = "invoice", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Set<Payment> payments;
}

File: apartment-resident-mobile/lib/features/invoices/models/invoice.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'invoice_item.dart';

part 'invoice.freezed.dart';
part 'invoice.g.dart';

enum InvoiceStatus {
  @JsonValue('UNPAID')
  unpaid,
  @JsonValue('PAID')
  paid,
  @JsonValue('OVERDUE')
  overdue,
}

@freezed
class InvoiceModel with _$InvoiceModel {
  const factory InvoiceModel({
    required int id,
    required int apartmentId,
    required String billingPeriod,
    required String issueDate,
    required String dueDate,
    required int totalAmount,
    required InvoiceStatus status,
    String? remarks,
    List<InvoiceItemModel>? items,
    String? createdAt,
    String? updatedAt,
  }) = _InvoiceModel;

  factory InvoiceModel.fromJson(Map<String, dynamic> json) =>
      _$InvoiceModelFromJson(json);
}
